backends/p4fmt/attach.h

```diff
@@ -0,0 +1,259 @@
+    }
+    void doneVec() {
+        BUG_CHECK(!vectorSeparator.empty(), "Empty vectorSeparator");
+        vectorSeparator.pop_back();
+    }
+    VecPrint getSep() {
+        BUG_CHECK(!vectorSeparator.empty(), "Empty vectorSeparator");
+        return vectorSeparator.back();
+    }
+
+    void doneList() {
+        BUG_CHECK(!listTerminators.empty(), "Empty listTerminators");
+        listTerminators.pop_back();
+    }
+    bool isSystemFile(cstring file);
+    cstring ifSystemFile(const IR::Node *node);  // return file containing node if system file
+    // dump node IR tree up to depth - in the form of a comment
+    void dump(unsigned depth, const IR::Node *node = nullptr, unsigned adjDepth = 0);
+    unsigned curDepth() const;
+
+ public:
+    // Output is constructed here
+    Util::SourceCodeBuilder &builder;
+    /* FIXME  -- simplify this by getting rid of the 'builder' object and just emitting
+     * directly to the ostream.  The SourceCodeBuilder object does not appear to add any
+     * useful functionality the ostream does not already provide; it just serves to
+     * obfuscate the code */
+    std::ostream *outStream;
+    /** If this is set to non-nullptr, some declarations
+        that come from libraries and models are not
+        emitted. */
+    cstring mainFile;
+
+    Attach(Util::SourceCodeBuilder &builder, bool showIR, cstring mainFile = nullptr)
+        : expressionPrecedence(DBPrint::Prec_Low),
+          isDeclaration(true),
+          showIR(showIR),
+          withinArgument(false),
+          builder(builder),
+          outStream(nullptr),
+          mainFile(mainFile) {
+        visitDagOnce = false;
+        setName("Attach");
+    }
+    Attach(std::ostream *outStream, bool showIR, cstring mainFile = nullptr)
+        : expressionPrecedence(DBPrint::Prec_Low),
+          isDeclaration(true),
+          showIR(showIR),
+          withinArgument(false),
+          builder(*new Util::SourceCodeBuilder()),
+          outStream(outStream),
+          mainFile(mainFile) {
+        visitDagOnce = false;
+        setName("Attach");
+    }
+    Attach()
+        :  // this is useful for debugging
+          expressionPrecedence(DBPrint::Prec_Low),
+          isDeclaration(true),
+          showIR(false),
+          withinArgument(false),
+          builder(*new Util::SourceCodeBuilder()),
+          outStream(&std::cout),
+          mainFile(nullptr) {
+        visitDagOnce = false;
+        setName("Attach");
+    }
+
+    using Inspector::preorder;
+
+    void setnoIncludesArg(bool condition) { noIncludes = condition; }
+
+    void setListTerm(const char *start, const char *end) {
+        listTerminators.push_back(ListPrint(start, end));
+    }
+    Visitor::profile_t init_apply(const IR::Node *node) override;
+    void end_apply(const IR::Node *node) override;
+
+    bool process(const IR::Type_StructLike *t, const char *name);
+    // types
+    bool preorder(const IR::Type_Boolean *t) override;
```

Comment:
How you can be sure the list is exhaustive? And that you won't miss something should new IR node kinds would be added?

Reply:
I think this is a valid concern. Do we have an answer to this?

Reply:
One way to make sure is to add a preorder for `IR::Node` and throw `P4C_UNIMPLEMENTED` once it is hit.